File: frontend-app/Baeckerei/model/BakedGood.swift
//
//  BakedGood.swift
//  Baeckerei
//
//  Created by Gregory Seibert on 02.02.19.
//  Copyright © 2019 Gregory Seibert. All rights reserved.
//

import Foundation

class BakedGood: Codable {
    let id: Int
    var name, characteristic: String
    var weight, kcal, fat, carbohydrates, protein: Double
    var pictureFilename: String
    
    var cerealMix: [CerealMixPercentage]
    var ingredients: [Ingredient]
    
    func hasAllergyType(allergyType: AllergyType) -> Bool {
        for ingredient in ingredients {
            if ingredient.allergyTypes.contains(allergyType) {
                return true
            }
        }
        
        for cerealMix in cerealMix {
            if cerealMix.ingredient.allergyTypes.contains(allergyType) {
                return true
            }
        }
        
        return false
    }
    
    func getWeightAsString() -> String {
        return "\(weight)g"
    }
    
    func getNutrientsAsString() -> String {
        return """
        \(kcal)\(NSLocalizedString("bakedgood_nutrients_kcal", comment: ""))
        \(protein)g \(NSLocalizedString("bakedgood_nutrients_protein", comment: ""))
        \(carbohydrates)g \(NSLocalizedString("bakedgood_nutrients_carbohydrates", comment: ""))
        \(fat)g \(NSLocalizedString("bakedgood_nutrients_fat", comment: ""))
        """
    }
    
    func getCerealMixAsString() -> String {
        return cerealMix.map{(cerealMix) -> String in
            return cerealMix.toString()
        }.joined(separator: ", ")
    }
    
    func getIngredientsAsString() -> String {
        return ingredients.map{(ingredient) -> String in
            return ingredient.toString()
        }.joined(separator: ", ")
    }
}

File: backend/src/main/java/de/gregoryseibert/baeckereibackend/model/BakedGood.java
package de.gregoryseibert.baeckereibackend.model;

import lombok.Data;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.List;

/**
 * The BakedGood class represents a baked good.
 *
 * @author Gregory Seibert
 * @version 1.0
 */

@Data
@MappedSuperclass
public abstract class BakedGood {
    @NotEmpty
    private String name;

    @Lob
    @NotEmpty
    private String characteristic;

    @NotNull
    private Double weight;

    @NotNull
    private Double kcal;

    @NotNull
    private Double fat;

    @NotNull
    private Double carbohydrates;

    @NotNull
    private Double protein;

    private String pictureFilename;

    @NotEmpty
    @OneToMany(cascade = CascadeType.ALL)
    private List<CerealMixPercentage> cerealMix;

    @NotEmpty
    @ManyToMany
    private List<Ingredient> ingredients;
}
